### PR TITLE
[core] Switch Async Callback to C++ [WIP]

### LINKED ISSUE #9251: Aysyncio ObjectID becomes None
### What is the problem?

Calling `await` on a task's ObjectID inside an async Actor sometimes results in the following error `AttributeError: 'NoneType' object has no attribute 'call_soon_threadsafe'`. This happens when the task runs after the actor method returns. This does not happen if `asyncio.wait()` is called. Sometimes this issue results in a weird segfault.

*Ray version and other system information (Python version, TensorFlow version, OS):*

### Reproduction (REQUIRED)
Please provide a script that can be run to reproduce the issue. The script should have **no external library dependencies** (i.e., use fake or mock data / environments):

- Change this [line](https://github.com/ray-project/ray/blob/master/python/ray/serve/router.py#L324) to `await worker.handle_request.remote(req)`
- Run `pytest -vs serve/tests/test_api.py::test_shadow_traffic`

If we cannot run your script, we cannot fix your issue.

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/latest/installation.html).

### PR DESCRIPTION
**Pull request by ijrsvt**
## Why are these changes needed?

The current implementation of Async functionality for ObjectIDs is confusing and spread out between C++ and Python. This PR moves as much logic as possible into C++. 

## Related issue number

Closes #9251
Supercedes #9161

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by edoakes** [comment]
@ijrsvt looks great! Could you make a full PR that removes the old code?

File: python/ray/experimental/async_api.py [hunk 1]
```diff
@@ -1,45 +1,10 @@
...
-    global handler
-    if handler is None:
-        worker = ray.worker.global_worker
-        loop = asyncio.get_event_loop()
-        handler = PlasmaEventHandler(loop, worker)
-        worker.core_worker.set_plasma_added_callback(handler)
-        logger.debug("AsyncPlasma Connection Created!")
-
-
 def as_future(object_id):
```
Comment (simon-mo) [inline python/ray/experimental/async_api.py:1]: - do we still need this? I think we can just remove the entire file right?

File: src/ray/core_worker/core_worker.h [hunk 2]
```diff
@@ -1065,8 +1065,15 @@ class CoreWorker : public rpc::CoreWorkerServiceHandler {
...
 
-  // Plasma Callback
-  PlasmaSubscriptionCallback plasma_done_callback_;
+  // Guard for Plasma Callback Map
+  mutable absl::Mutex plasma_mutex_;
+
+  // Plasma Callbacks
+  absl::flat_hash_map<ObjectID, std::vector<std::function<void(void)>>>
+      async_plasma_callbacks_ GUARDED_BY(plasma_mutex_);
+
```
Comment (simon-mo) [inline src/ray/core_worker/core_worker.h:1074]: document this?

File: src/ray/core_worker/core_worker.h [hunk 3]
```diff
@@ -759,14 +759,14 @@ class CoreWorker : public rpc::CoreWorkerServiceHandler {
   void GetAsync(const ObjectID &object_id, SetResultCallback success_callback,
                 SetResultCallback fallback_callback, void *python_future);
 
-  /// Connect to plasma store for async futures
-  using PlasmaSubscriptionCallback = std::function<void(ray::ObjectID, int64_t, int64_t)>;
-
-  /// Set callback when an item is added to the plasma store.
+  /// Perform async get from in-memory store.
```
Comment (simon-mo) [inline src/ray/core_worker/core_worker.h:762]: this also does retrying?

**Review by simon-mo — COMMENTED** [review]
This is awesome!!!

File: src/ray/core_worker/core_worker.h (hunk 2 above)
Comment (simon-mo) [reply src/ray/core_worker/core_worker.h:1074]: misplaced, this comment is about documenting PlasmaCallback method

File: src/ray/core_worker/core_worker.cc [hunk 4]
```diff
@@ -1999,8 +1999,53 @@ void CoreWorker::GetAsync(const ObjectID &object_id, SetResultCallback success_c
...
+      // example if the callback calls ray.get() on an object that is dependent on an RPC
+      // to be ready).
+      callback();
+    }
+  }
+}
+
+void CoreWorker::GetAsyncNew(const ObjectID &object_id,
+                             SetResultCallback success_callback, void *python_future) {
+  GetAsync(object_id, success_callback,
```
[the remark this answers is not in the document]
Comment (ijrsvt) [reply src/ray/core_worker/core_worker.cc:2045]: Definitely!

File: python/ray/experimental/async_api.py (hunk 1 above)
Comment (ijrsvt) [reply python/ray/experimental/async_api.py:1]: Yes we can!

File: src/ray/core_worker/core_worker.h (hunk 3 above)
Comment (ijrsvt) [reply src/ray/core_worker/core_worker.h:762]: What do you mean? in terms of fetching from Plasma?

File: src/ray/core_worker/core_worker.h (hunk 2 above)
Comment (ijrsvt) [reply src/ray/core_worker/core_worker.h:1074]: I'll modify the comment to be more explanative.

File: python/ray/experimental/test/async_test.py [hunk 5]
```diff
@@ -1,20 +1,18 @@
 import asyncio
```
Comment (simon-mo) [inline python/ray/experimental/test/async_test.py:1]: can we move this test to python/ray/tests/test_async?

File: src/ray/core_worker/core_worker.h [hunk 6]
```diff
@@ -1065,8 +1055,15 @@ class CoreWorker : public rpc::CoreWorkerServiceHandler {
...
-  // Plasma Callback
-  PlasmaSubscriptionCallback plasma_done_callback_;
+  // Guard for `async_plasma_callbacks_` map.
+  mutable absl::Mutex plasma_mutex_;
+
+  // Callbacks for when when a plasma object becomes ready.
+  absl::flat_hash_map<ObjectID, std::vector<std::function<void(void)>>>
+      async_plasma_callbacks_ GUARDED_BY(plasma_mutex_);
+
+  void PlasmaCallback(SetResultCallback success, std::shared_ptr<RayObject> ray_object,
```
Comment (simon-mo) [inline src/ray/core_worker/core_worker.h:1065]: PlasmaCallback needs a comment

File: src/ray/core_worker/core_worker.cc [hunk 7]
```diff
@@ -1987,8 +1987,61 @@ void CoreWorker::YieldCurrentFiber(FiberEvent &event) {
...
+}
+
+/**
+ *
+ * void CoreWorker::GetAsync(const ObjectID &object_id, SetResultCallback
+success_callback, SetResultCallback fallback_callback, void *python_future) {
+
+}
+**/
+
```
Comment (simon-mo) [inline src/ray/core_worker/core_worker.cc:2038]: remove

File: src/ray/core_worker/core_worker.h (hunk 6 above)
Comment (simon-mo) [reply src/ray/core_worker/core_worker.h:1065]: Just to confirm, this handler is for GetAsync falling back to plasma?

File: src/ray/core_worker/core_worker.cc [hunk 8]
```diff
@@ -1987,8 +1987,61 @@ void CoreWorker::YieldCurrentFiber(FiberEvent &event) {
   event.Wait();
 }
 
+void CoreWorker::PlasmaCallback(SetResultCallback success,
+                                std::shared_ptr<RayObject> ray_object, ObjectID object_id,
+                                void *py_future) {
+  std::vector<std::shared_ptr<RayObject>> vec;
+  if (Get(std::vector<ObjectID>{object_id}, 0, &vec).ok() && vec.size() > 0) {
+    return success(vec.front(), object_id, py_future);
+  }
```
Comment (simon-mo) [inline src/ray/core_worker/core_worker.cc:1996]: what's this case for?

**Review by simon-mo — APPROVED** [review]

File: python/ray/experimental/test/async_test.py (hunk 5 above)
Comment (ijrsvt) [reply python/ray/experimental/test/async_test.py:1]: Sure!

File: src/ray/core_worker/core_worker.h (hunk 6 above)
Comment (ijrsvt) [reply src/ray/core_worker/core_worker.h:1065]: Yep!

File: src/ray/core_worker/core_worker.cc (hunk 8 above)
Comment (ijrsvt) [reply src/ray/core_worker/core_worker.cc:1996]: Any case where the object is immediately available, but not in Memory. Calling before subscribe avoids filling subscribe with unnecessary object ids.

**Pull request merged**